### ISSUE TITLE
LTC job completion delayed by one or two slow machines

### ISSUE THREAD

**Issue by xoto10** [glinscott/fishtest#131]
Job completion often seems to be held up by one or two slow machines - fishtest allocates 10 machines to do the last 10x1000 games (thinking of NumGames or SPSA runs here), and obviously the single core machines take longer than the multi-core machines. In terms of time delay this affects longer time controls most, it's not such a big deal for STC. A couple of ideas that might reduce this:
Perhaps single core machines could only be allocated to STC jobs, this would reduce this effect on long running jobs?
For NumGames and SPSA runs, maybe the last 10000 games could be allocated in half sizes, i.e. 500 games per machine instead of 1000? This would also reduce the runtime difference caused by differing machines.

**Comment by Stefano80** [comment]
By is this an issue?

**Comment by xoto10** [comment]
Well, I'm not sure. The overall throughput wouldn't change, so maybe it's not important. (I think. maybe.) If we're effectively just slightly re-allocating from newer to older jobs then on average nothing much changes, jobs just start off slightly slower and then finish a bit quicker. 
But it does seem a bit frustrating to have to wait for one last slow worker, (possibly for quite a while), if you want the "final" result of your test.

**Comment by Stefano80** [comment]
Well than stop the test and call the result final. Or change the number of games.

May I close this?

**Comment by xoto10** [comment]
Well, it's your decision. 
For an example, look at http://tests.stockfishchess.org/tests/view/5a60b2820ebc590297b9b7e0
Ok, VLTC is unusual, but 3 workers have finished now, while 2 of them will take weeks at the current rate of progress. LTC tests face a similar issue, but not quite as pronounced. Rearranging allocation slightly to get the same throughput but quicker finishes on LTC tests seems like an easy win to me.
I'm just making suggestions, if you want to close it, close it.

**Comment by Stefano80** [comment]
In my opinion this is not an issue. I think we should keep this issue list reserved to problems which affect contributors or to bugs/errors affecting SF development.

I nevertheless think this could be a useful improvement, if you have some simple idea to implement this, just try it out.

I will keep it open for some days if any people feels like contributing and then I will close it.

**Comment by xoto10** [comment]
Ok, that seems fair.

**Comment by xoto10** [comment]
I've attempted a couple of code changes, see https://github.com/xoto10/fishtest/tree/quickerltc1 and https://github.com/xoto10/fishtest/tree/quickerltc2 but I'm still in the process of setting up a local fishtest so I don't know if they work yet. (I notice there's a comma missing in the first one for a start.)

**Comment by ppigazzini** [comment]
@xoto10 the wiki has a setup script to install a [fishtest server on Ubuntu](https://github.com/glinscott/fishtest/wiki/Installing-a-development-fishtest-server-on-Ubuntu). Let me know if you need a script for CentOS.

**Comment by xoto10** [comment]
Yeah, I followed that one. I had some trouble networking the vm but got it going and was starting to work on the chnages ... then I did a big os update and it blew up the os somehow :-(
I've reinstalled ubuntu and tomorrow I'll set up fishtest again ... and I'll back up the vm this time!

I got as far as creating a couple of tests today, but the worker said there was no work to do? Let me know if you have any idea what could cause that. I commented out the lines checking the number of cores (currently only 1 core in the vm ... another thing to fix ...) but got a message saying no jobs available or something like that. I'll carry on during the week ...

**Comment by ppigazzini** [comment]
To be sure: you have created a couple of tests w/ "user01" and then you have approved those tests w/ "user00", correct?

**Comment by xoto10** [comment]
Yes, the 2 tests were "Pending". It probably didn't help that I started the worker in the vm, there was no need to do that. When I've got it going again I'll start the worker outside the vm this time ... and I won't update any code until I've seen one or two tests work either ...

**Comment by ppigazzini** [comment]
@xoto10 just to be sure to be on the same page: a test is created in status "pending" but it must be approved.
Check the second icon from the left: if it is orange w/ an interrogative point, the test must be still approved. W/ an approved test the icon is green with a thumb.

Anyway I just installed a fishtest server within a vmware VM using this [updated setup script for Ubuntu](https://gist.github.com/ppigazzini/0c7968edd80929a8d4ad94345177e460), created and approved 3 tests, installed the packages required by a [linux worker](https://github.com/glinscott/fishtest/wiki/Running-the-worker-on-Linux), started a couple of workers inside the VM. At the moment all is working fine.

**Comment by xoto10** [comment]
Yes, sure. I have it working now - thanks for your help!

**Comment by ppigazzini** [comment]
@xoto10 great 👍

**Comment by xoto10** [comment]
OK. I think these code changes work, and would be useful for finishing the fixed-length LTC/VLTC jobs a bit quicker:
https://github.com/glinscott/fishtest/compare/master...xoto10:quickerltc1
https://github.com/glinscott/fishtest/compare/master...xoto10:quickerltc2
ltc1 means workers with less than 3 cores can only work on 10+0.1 tests
ltc2 reduces the task size from 1000 to 500 (250 for VLTC) when in the last 10000 of a fixed size test

I'll let you guys take it from here if you want to, I think these would be useful changes to allow longer regression / tuning jobs to finish a bit sooner.

**Comment by xoto10** [comment]
Hmmm, some clarification:
ltc1 does what it does so I think that's good
ltc2 relies on a test getting workers gradually allocated to it over time. This is not so great, since one example of the problem I'm trying to address is the current regression test http://tests.stockfishchess.org/tests/view/5a673c8a0ebc590d945d5815
Because many workers were free when this was created, it got a full complement of 40 workers allocated straightaway to do 40x1000 games, so reducing the size of the tasks for the last 10000 games wouldn't help. It still may be useful, but perhaps more thought is required on this one.

**Comment by Stefano80** [comment]
Hi xoto, take into account the case in which there are only test available for which the worker is not allowed. I think the only really foolproof method would be to make the internal priority worker-dependent, but again this would be a bit invasive.

**Comment by xoto10** [comment]
Yes, I thought of that. Looking at the workers, there aren't many 1 or 2 core workers, and as a % of the cores available obviously it's an even smaller number. And I assume there is a trend towards more cores over time, so the number of 1 or 2 core workers is more likely to decrease further than increase. The thing is, a single core worker just doesn't do much on a large LTC job, so I don't think missing them out altogether is a problem.

**Comment by Stefano80** [comment]
Well, if the patch is supposed to improve efficiency and we end up losing workers and increasing time spent compiling, plus additionally have a more complicated code, then I would prefer to pass.

But if you can find a simple solution for worker dependent internal priority, then this could make sense.

**Comment by xoto10** [comment]
On the 3 core worker thing, I guess I could make it loop round (or do a second query) if no job is found, so a 1-core worker would only get an LTC test if there was nothing else available, but that seems a bit clunky. Let me look into what's available in mongodb a bit more ...

**Comment by Stefano80** [comment]
I'll leave this open for some be time.

**Comment by xoto10** [comment]
Ok, I have two versions of the 3-core worker idea.
- As above, only allow 1 and 2 core workers to work on 10+0.1 jobs: https://github.com/glinscott/fishtest/compare/master...xoto10:quickerltc1
- Modified version, make 1 and 2 core workers take a 10+0.1 job if there is one, otherwise take any job: https://github.com/xoto10/fishtest/commit/e74c1b893a1638ca819b940162b9752c17e2c739

My preference is not to give long jobs to 1 or 2 core workers at all, the current regression test http://tests.stockfishchess.org/tests/view/5a673c8a0ebc590d945d5815 is a good example. The smaller workers only do perhaps 10% of the work on a job like this, so removing them is not a big problem, and it removes the effect where the last 2/3000 games take ages to finish because of these slow workers. If no slow workers had been assigned to this test it would have been finished ages ago by a faster worker. The second option probably achieves this some of the time, with the advantage that if there are no 10+0.1 jobs waiting, the slower workers will get allocated to something.

**Comment by xoto10** [comment]
I have a suggestion for improving the task information. Parsing the tc is pretty tricky and I see there is a special function to do this; parse_tc(). I think it would be useful if the result of this was added into the task info, e.g. line 39 in rundb.py
```      tasks.append({
        'num_games': task_size,
        'pending': True,
        'active': False,
+        'game_time': parse_tc(tc),
      })
```
I don't understand mongodb enough yet to know if you can just add that, or if that means a new database structure and a major upgrade ... something to consider for the future though. If the expected game_time in seconds was in the task info that might make better task allocation possible in the future ...

**Comment by Stefano80** [comment]
It looks complicated, give some days to get through the logics

**Comment by xoto10** [comment]
Sure, no worries.

That link I gave above wasn't very good. I'm sure you can work github better than me but I'll correct it anyway, I should have said:

Ok, I have two versions of the 3-core worker idea.
- As above, only allow 1 and 2 core workers to work on 10+0.1 jobs: https://github.com/glinscott/fishtest/compare/master...xoto10:quickerltc1
- Modified version, make 1 and 2 core workers take a 10+0.1 job if there is one, otherwise take any job: https://github.com/glinscott/fishtest/compare/master...xoto10:quickerltc1b [better link]

...

**Comment by d3vv** [comment]
And what about 1-core workers? Kill'EM ALL? about dew-machines? ))

**Comment by Stefano80** [comment]
Hi @xoto10, I've gone through your proposal and I think it is correctly and cleanly coded, thank you for the effort. Nevertheless, I thought through the issue and I would to close the issue and not merge your proposal based on the following reasoning.

One main goals for us must be to keep fishtest efficiency at its maximum. This means: for a given number of machines logged to the framework, we have to squeeze the maximum number of games per hour (or better: computed nodes per hour).

This means, among others, to reduce the fraction of machine time which is spent at compiling and increase the fraction spent with playing games. As an example, my machine goes 1h30' per 1000 games batch and needs almost exactly 2 minutes for 2x profile-building. Which means the time spent in compiling is ca. 2%.

In particular, this means that the optimal configuration is that every machine runs all its batches to end. Your change would allocate slow workers to fast finishing tests, reducing the probability that they can finish the batch.

I know this is a small effect, but I would'nt like to add code which slows down the framework for purely aesthetic reasons.

I hope this does not discourage you from further contributions and I really valued the exchange with you and your accuracy.

**Comment by Stefano80** [comment]
And sorry for taking so much time, I was somewhat taken with my dynamic contempt experiments.

**Comment by xoto10** [comment]
@Stefano80 No worries about the time, this is clearly a lower priority, longer term type thing. I was/am much more interested in your dynamic contempt work as well. In fact I nearly submitted a threaded version of your work today, but @WOnder93 beat me to it.

**Comment by xoto10** [comment]
dew / @d3vv hope you didn't take offence at my patch! :-)
Thank you for contributing so many machines to fishtest, the 1 core ones and the bigger ones! :-)
